File: 2022/day2/day2.clj
(require '[clojure.string :as str]
         '[clojure.set :as set])

(def input-raw (slurp "2022/day2/input.txt"))
#_(def input-raw "
A Y
B X
C Z
")

(def input
  (->> input-raw
       str/trim
       str/split-lines
       (map (fn [line] (str/split line #" ")))))

(defn index-of [x xs]
  (first (keep-indexed #(when (= %2 x) %1) xs)))

(def youList ["A" "B" "C"])
(def meList ["X" "Y" "Z"])
(def win-conditions {"Y" "A" "Z" "B" "X" "C"})
(def lose-conditions {"X" "B" "Y" "C" "Z" "A"})
(def tie-conditions {"X" "A" "Y" "B" "Z" "C"})

(defn score-round [you me]
  (+ (cond
       (= (win-conditions me) you) 6
       (= (lose-conditions me) you) 0
       (= (tie-conditions me) you) 3)
     (inc (index-of me meList))))

(defn part1 [input]
  (->> input
       (map #(apply score-round %))
       (apply +)))

(defn part2 [input]
  (let [choose (fn [you strategy] 
                 (case strategy
                   "X" ((set/map-invert lose-conditions) you)
                   "Y" ((set/map-invert tie-conditions) you)
                   "Z" ((set/map-invert win-conditions) you)))
        round-result (fn [you strategy]
                       [you (choose you strategy)])]
    (->> input
         (map #(apply round-result %))
         (map #(apply score-round %))
         (apply +)
         )))

(prn "part1:" (part1 input))
(prn "part2:" (part2 input))

File: 2022/day2/day2.js
const fs = require('fs');
const path = require('path');

function part1(input) {
  /** @type string[] */
  let lines = input.trim().split('\n');
  let youMap = ['A', 'B', 'C'];
  let meMap = ['X', 'Y', 'Z'];

  let score = 0;
  for (let line of lines) {
    let [you, me] = line.split(' ');
    let youIndex = youMap.indexOf(you);
    let meIndex = meMap.indexOf(me);
    let result = 0;
    if (meIndex === youIndex) {
      result = 3;
    } else if (youIndex === 0 && meIndex === 2) {
      result = 0;
    } else if (meIndex > youIndex || (youIndex === 2 && meIndex === 0)) {
      result = 6;
    }
    score += meIndex + 1 + result;
  }
  return score;
}

function part2(input) {
  /** @type string[] */
  let lines = input.trim().split('\n');
  let youMap = ['A', 'B', 'C'];
  let meMap = ['X', 'Y', 'Z'];

  let score = 0;
  for (let line of lines) {
    let [you, me] = line.split(' ');
    let youIndex = youMap.indexOf(you);
    let meIndex;
    if (me === 'Y') {
      // draw
      meIndex = youIndex;
    } else if (me === 'X') {
      // lose
      meIndex = (youIndex - 1 + 3) % 3;
    } else {
      // win
      meIndex = (youIndex + 1 + 3) % 3;
    }
    let result = 0;
    if (meIndex === youIndex) {
      result = 3;
    } else if (youIndex === 0 && meIndex === 2) {
      result = 0;
    } else if (meIndex > youIndex || (youIndex === 2 && meIndex === 0)) {
      result = 6;
    }
    score += meIndex + 1 + result;
  }
  return score;
}

let input = fs.readFileSync(path.resolve(__dirname, './input.txt'), 'utf8').replaceAll('\r', '');

// input = `
// A Y
// B X
// C Z`;

console.log('part1:', part1(input));
console.log('part2:', part2(input));
